src/CodeMirrorBlocks.tsx

```diff
@@ -24,20 +27,68 @@ export type Options = {
   incrementalRendering?: boolean;
 };
 
+/**
+ * A language definition object
+ */
 export type Language = {
+  /**
+   * A unique id for the language
+   */
   id: string;
+
+  /**
+   * The name of the language
+   */
   name: string;
-  parse: () => void;
-  getExceptionMessage?: () => void;
-  getASTNodeForPrimitive?: () => void;
-  getLiteralNodeForPrimitive?: () => void;
-  primitivesFn?: () => unknown[];
+
+  /**
+   * A function for generating an AST from source code
+   * @param code source code for the program
+   * @returns The ast that codemirror-blocks will render
+   */
+  parse(code: string): AST;
+
+  /**
+   * A function for generating a human readable error message
+   * from any exceptions that are thrown by a call to parse()
+   */
+  getExceptionMessage?(e: any): string;
+
+  /**
+   * A function for generating ASTNodes from Primitives
+   */
+  getASTNodeForPrimitive?: (primitive: Primitive) => ASTNode;
+
+  /**
+   * A function for generating a Literal ast node from a Primitive
+   */
+  getLiteralNodeForPrimitive?: (primitive: Primitive) => Literal;
+
+  /**
+   * Returns a list of language primitives that will be displayed
+   * in the search bar.
+   */
+  primitivesFn?: () => Primitive[];
+
+  /**
+   * @deprecated
+   */
+  primitives: Primitive[];
```

Comment:
The handling of `primitivesFn` and `primitives` seems really weird. The language modules that implement `primitivesFn` also set `primitives` when that function is called, in addition to returning the primitives that get generated. So I marked `primitives` as deprecated since it seems unnecessary to have both. It's not obvious to me what was originally intended.

Reply:
I suspect this is a holdover from some work in 2019 that was designed to let users change language without having to reload the entire editor. What was once a static property now became dynamic. Deprecation is the way to go here, and nowhere else in the codebase should use it anyway!